scripts/install-cni.sh

```diff
@@ -130,20 +130,149 @@ else
+
+  cni_spec=${cni_spec//@subnets/$SUBNETS_REPLACEMENT_CONCATENATED}
+  cni_spec=${cni_spec//@routes/$ROUTES_REPLACEMENT_CONCATENATED}
+}
+
+#######################################
+# Replaces `@ipv4Subnet', '@ipv6SubnetOptional` and `@ipv6RouteOptional` placeholders using `.spec.podCIDR` from node json and node's ipv6 from metadata server.
+# Arguments:
+#   node json from kube-apiserver
+#   node_ipv6_addr node's IPv6 address from GCE metadata server
+#######################################
+function fillSubnetsInCniSpecLegacyTemplate {
+  local node=$1
+  local node_ipv6_addr=$2
+
+  local primary_subnet
+  primary_subnet=$(jq -r '.spec.podCIDR' <<<"$response")
+
+  if is_ipv4_range "${primary_subnet:-}" ; then
+    echo "PodCIDR validation succeeded: ${primary_subnet:-}"
+    echo "Filling IPv4 subnet ${primary_subnet:-}"
+    cni_spec=${cni_spec//@ipv4Subnet/[{\"subnet\": \"${primary_subnet:-}\"\}]}
+  elif is_ipv6_range "${primary_subnet:-}" ; then
+    echo "Primary IPv6 pod range detected '${primary_subnet:-}'. It will only work with new spec template."
+  else
+    echo "Response from $node_url"
+    echo "$response"
+    echo "Failed to fetch/validate PodCIDR from K8s API server, primary_subnet=${primary_subnet:-}. Exiting (1)..."
+    exit 1
+  fi
+
+  if [ "$ENABLE_IPV6" == "true" ] || [ "${STACK_TYPE:-}" == "IPV4_IPV6" ]; then
+    if [ -n "${node_ipv6_addr:-}" ] && [ "${node_ipv6_addr}" != "null" ]; then
+      echo "Found nic0 IPv6 address ${node_ipv6_addr:-}. Filling IPv6 subnet and route..."
+      POPULATE_IP6TABLES="true"
+
+      cni_spec=${cni_spec//@ipv6SubnetOptional/, [{\"subnet\": \"${node_ipv6_addr:-}/112\"\}]}
+      cni_spec=${cni_spec//@ipv6RouteOptional/, ${CNI_SPEC_IPV6_ROUTE:-{\"dst\": \"::/0\"\}}}
+    else
+      echo "No IPv6 address found for nic0. Clearing IPv6 subnet and route..."
+      cni_spec=${cni_spec//@ipv6SubnetOptional/}
+      cni_spec=${cni_spec//@ipv6RouteOptional/}
+    fi
+  else
+    echo "Clearing IPv6 subnet and route given IPv6 access is disabled..."
+    cni_spec=${cni_spec//@ipv6SubnetOptional/}
+    cni_spec=${cni_spec//@ipv6RouteOptional/}
+  fi
+
+}
+
+function fillSubnetsInCniSpec {
+  fillSubnetsInCniSpecV2Template "$1" "$2"
+  fillSubnetsInCniSpecLegacyTemplate "$1" "$2"
+}
+
+
+STACK_TYPE=$(jq -r '.metadata.labels."cloud.google.com/gke-stack-type"' <<<"$response")
+echo "Node stack type label: '${STACK_TYPE:-}'"
+
+node_ipv6_addr=''
+if [ "$ENABLE_IPV6" == "true" ] || [ "${STACK_TYPE:-}" == "IPV4_IPV6" ]; then
+  node_ipv6_addr=$(curl -s -k --fail "http://metadata.google.internal/computeMetadata/v1/instance/network-interfaces/0/?recursive=true" -H "Metadata-Flavor: Google" | jq -r '.ipv6s[0]' ) ||:
 fi
 
-echo "Filling IPv4 subnet ${ipv4_subnet:-}"
-cni_spec=${cni_spec//@ipv4Subnet/[{\"subnet\": ${ipv4_subnet:-}\}]}
+fillSubnetsInCniSpec "$response" "$node_ipv6_addr"
 
 if [ "$ENABLE_MASQUERADE" == "true" ]; then
```

Comment:
By the way I think this chunk can be removed... @MrHohn to verify?

Reply:
Oh I missed this - yes this is legacy logic that is no longer in used. Don't have to do it in the same PR though.

Reply:
Yeah just spotted it here since this has some major overhaul of the script.